### PR TITLE
Feature/new-dipole-extern

### PR DESCRIPTION
**Pull request by xiyudu**
## Description

This is a draft pull request to port anisotropic pair potentials over to HOOMD 3.0.  The goal here is to port pair.dipole and pair.gb over, and then create an external dipole field.  Right now, the two anisotropic potentials are ported over and I am working on the external field.

## Motivation and Context

This pull request is to help port over anisotropic potentials to HOOMD 3.0 and also provide a test bed to later port over other anisotropic potentials for janus and patchy particles.
## How Has This Been Tested?

Unit test is under development.

## Change log

```

```

## Checklist:

- [ ] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/hoomd-blue/blob/master/CONTRIBUTING.md).
- [ ] I agree with the terms of the [**HOOMD-blue Contributor Agreement**](https://github.com/glotzerlab/hoomd-blue/blob/master/ContributorAgreement.md).
- [ ] My name is on the [list of contributors](https://github.com/glotzerlab/hoomd-blue/blob/master/sphinx-doc/credits.rst).

### CODE REVIEW

File: hoomd/md/AnisoPotentialPair.h [hunk 1]
```diff
@@ -379,6 +442,33 @@ void AnisoPotentialPair< aniso_evaluator >::setParams(unsigned int typ1, unsigne
...
+    {
+    // TODO change logic to just throw an exception
+    auto n_types = this->m_pdata->getNTypes();
+    if (typ1 >= n_types || typ2 >= n_types)
+        {
+        this->m_exec_conf->msg->error() << "pair." << aniso_evaluator::getName()
+            << ": Trying to " << action << " for a non existent type! "
+            << typ1 << "," << typ2 << std::endl;
+        throw std::runtime_error("Error setting parameters in AnisoPotentialPair");
+        }
```
Comment (b-butler) [inline hoomd/md/AnisoPotentialPair.h:469]: This can be changed to just throw an exception now.

File: hoomd/md/EvaluatorPairDipole.h [hunk 2]
```diff
@@ -94,7 +78,55 @@ struct dipole_shape_params
...
+            #ifndef __HIPCC__
+            param_type() {mu = 0; A = 0; kappa = 0;}
+
+            // this constructor facilitates unit testing
+            //param_type(Scalar muu, Scalar Aa, Scalar kappaa)
+            //    {
+            //    mu = muu;
+            //    A = Aa;
+            //    kappa = kappaa;
+            //    }
```
Comment (b-butler) [inline hoomd/md/EvaluatorPairDipole.h:104]: Why is this commented out?

File: hoomd/md/EvaluatorPairDipole.h [hunk 3]
```diff
@@ -288,17 +320,10 @@ class EvaluatorPairDipole
         Scalar rcutsq;              //!< Stored rcutsq from the constructor
         Scalar q_i, q_j;            //!< Stored particle charges
         Scalar4 quat_i,quat_j;      //!< Stored quaternion of ith and jth particle from constructor
-        const param_type &params;   //!< The pair potential parameters
+        Scalar mu;
+        Scalar A;
+        Scalar kappa;
```
Comment (b-butler) [inline hoomd/md/EvaluatorPairDipole.h:325]: Is there a reason you switched from storing the params in the `param_type`? Either way works, but I am just wondering why the change.

File: hoomd/md/EvaluatorPairGB.h [hunk 4]
```diff
@@ -74,7 +57,47 @@ struct gb_shape_params
...
+            #endif
+
+            #ifndef __HIPCC__
+            param_type() {epsilon = 0; lperp = 0; lpar = 0;}
+
+            param_type(pybind11::dict v)
+                {
+                auto epsilon(v["epsilon"].cast<Scalar>());
+                auto lperp(v["lperp"].cast<Scalar>());
+                auto lpar(v["lpar"].cast<Scalar>());
```
Comment (b-butler) [inline hoomd/md/EvaluatorPairGB.h:81]: This will not set `epsilon`, `lperp` or `lpar`. `epsilon = v["epsilon"].cast<Scalar>();`

File: hoomd/md/module-md.cc [hunk 5]
```diff
@@ -155,11 +155,11 @@ void export_pair_params(py::module& m)
     py::class_<pair_fourier_params>(m, "pair_fourier_params").def(py::init<>());
     m.def("make_pair_fourier_params", &make_pair_fourier_params);
 
-    py::class_<pair_dipole_params>(m, "pair_dipole_params").def(py::init<>());
-    m.def("make_pair_dipole_params", &make_pair_dipole_params);
+    //py::class_<pair_dipole_params>(m, "pair_dipole_params").def(py::init<>());
+    //m.def("make_pair_dipole_params", &make_pair_dipole_params);
```
Comment (b-butler) [inline hoomd/md/module-md.cc:159]: Don't comment out if they are no longer used, just remove the code.

File: hoomd/md/module-md.cc [hunk 6]
```diff
@@ -155,11 +155,11 @@ void export_pair_params(py::module& m)
...
 
-    py::class_<pair_gb_params>(m, "pair_gb_params").def(py::init<>());
-    m.def("make_pair_gb_params", &make_pair_gb_params);
+    //py::class_<pair_gb_params>(m, "pair_gb_params").def(py::init<>());
+    //m.def("make_pair_gb_params", &make_pair_gb_params);
```
Comment (b-butler) [inline hoomd/md/module-md.cc:162]: See above comment.

File: hoomd/md/pair.py [hunk 7]
```diff
@@ -2055,9 +1986,9 @@ def get_type_shapes(self):
         Returns:
             A list of dictionaries, one for each particle type in the system.
         """
-        return super(ai_pair, self)._return_type_shapes();
+        #return super(ai_pair, self)._return_type_shapes();
```
Comment (b-butler) [inline hoomd/md/pair.py:1989]: Why is this commented out?

File: hoomd/md/pair.py [hunk 8]
```diff
@@ -1866,50 +1866,24 @@ class ai_pair(pair):
...
+                              TypeParameterDict(r_cut, len_keys=2)
+                             )
+        r_on = TypeParameter('r_on', 'particle_types',
+                             TypeParameterDict(float(r_on), len_keys=2)
+                            )
+        self._extend_typeparam([r_cut, r_on])
+        self._param_dict.update(
+            ParameterDict(mode=OnlyFrom(['none', 'shifted', 'xplor']),
+                          explicit_defaults=dict(mode=mode))
+            )
```
Comment (b-butler) [inline hoomd/md/pair.py:1883]: This is identical to `_Pair.__init__`. No need to have this code.

File: hoomd/md/pair.py [hunk 9]
```diff
@@ -1866,50 +1866,24 @@ class ai_pair(pair):
...
-        if mode is not None:
-            if mode == "no_shift":
-                self.cpp_force.setShiftMode(self.cpp_class.energyShiftMode.no_shift)
-            elif mode == "shift":
-                self.cpp_force.setShiftMode(self.cpp_class.energyShiftMode.shift)
-            else:
-                hoomd.context.current.device.cpp_msg.error("Invalid mode\n");
-                raise RuntimeError("Error changing parameters in pair force");
-
     @property
```
Comment (b-butler) [inline hoomd/md/pair.py:1887]: We should remove this. It should be taken over by a `typeparameter`.

File: hoomd/md/pair.py [hunk 10]
```diff
@@ -1922,40 +1896,18 @@ def shape(self):
...
 
-        This method must be implemented by subclasses to generate the
-        appropriate shape structure. The default behavior is to do nothing."""
-        pass
+    @nlist.setter
+    def nlist(self, value):
+        if self.is_attached:
+            raise RuntimeError("nlist cannot be set after attaching.")
+        else:
+            self._nlist = validate_nlist(_NList)(value)
```
Comment (b-butler) [inline hoomd/md/pair.py:1908]: Remove please, already in `_Pair`.

File: hoomd/md/AnisoPotentialPair.h [hunk 11]
```diff
@@ -379,6 +442,33 @@ void AnisoPotentialPair< aniso_evaluator >::setParams(unsigned int typ1, unsigne
     h_params.data[m_typpair_idx(typ2, typ1)] = param;
     }
 
+template< class aniso_evaluator >
+void AnisoPotentialPair< aniso_evaluator >::setParamsPython(pybind11::tuple typ, pybind11::dict params)
```
Comment (b-butler) [inline hoomd/md/AnisoPotentialPair.h:446]: This isn't implemented in the child classes, and means we cannot actually set parameters in C++ from Python. Since the param_type of `AnisoPotentialPair` should always have a constructor from a Python `dict` we can just use that here.

Comment (b-butler) [reply hoomd/md/AnisoPotentialPair.h:446]: If that prevents testing from building which it probably will, then you will need to implement it for all child classes until all anisotropic pair potentials are converted.

File: hoomd/md/AnisoPotentialPair.h [hunk 12]
```diff
@@ -379,6 +442,33 @@ void AnisoPotentialPair< aniso_evaluator >::setParams(unsigned int typ1, unsigne
...
+    {}
+
+template< class aniso_evaluator >
+pybind11::dict AnisoPotentialPair< aniso_evaluator >::getParams(pybind11::tuple typ)
+    {
+    pybind11::dict v;
+    return v;
+    }
```
Comment (b-butler) [inline hoomd/md/AnisoPotentialPair.h:454]: Similar to above comment.

File: hoomd/md/pair.py (hunk 8 above)
Comment (b-butler) [reply hoomd/md/pair.py:1883]: This is also where the error in initializing you were getting comes from. The argument should be `_defaults` now not `explicit_defaults`.

**Comment by b-butler** [comment]
@joaander this is ready for review now.

File: hoomd/md/EvaluatorPairDipole.h [hunk 13]
```diff
@@ -187,19 +225,19 @@ class EvaluatorPairDipole
             Scalar r5inv = r3inv*r2inv;
 
             // convert dipole vector in the body frame of each particle to space frame
-            vec3<Scalar> p_i = rotate(quat<Scalar>(quat_i), vec3<Scalar>(params.mu, 0, 0));
-            vec3<Scalar> p_j = rotate(quat<Scalar>(quat_j), vec3<Scalar>(params.mu, 0, 0));
+            vec3<Scalar> p_i = rotate(quat<Scalar>(quat_i), vec3<Scalar>(mu, 0, 0));
```
Comment (joaander) [inline hoomd/md/EvaluatorPairDipole.h:228]: We could leave this for a later PR, but I'd like to generalize the potential to allow vector mu's per particle. We can use the `shape` data to store a vector mu per particle type, but expose that to the user as a tuple type parameter:
```
dipole.mu['A'] = (x,y,z)
```

**Review by joaander — CHANGES_REQUESTED** [review]
Thanks! I have a few minor suggestions.

File: hoomd/md/pair.py [hunk 14]
```diff
@@ -1641,127 +1647,29 @@ def __init__(self, nlist, r_cut=None, r_on=0., mode='none'):
...
 
-    Users should not instantiate :py:class:`ai_pair` directly. It is a base class that
-    provides common features to all anisotropic pair forces. Rather than repeating all of that documentation in a
-    dozen different places, it is collected here.
-
-    All anisotropic pair potential commands specify that a given potential energy, force and torque be computed
-    on all non-excluded particle pairs in the system within a short range cutoff distance :math:`r_{\mathrm{cut}}`.
-    The interaction energy, forces and torque depend on the inter-particle separation
-    :math:`\vec r` and on the orientations :math:`\vec q_i`, :math:`q_j`, of the particles.
+    Users should not instantiate :py:class:`ai_pair` directly. It is a base
```
[the remark this answers is not in the document]
Comment (b-butler) [reply hoomd/md/pair.py:1653]: I was thinking that there were no specific details of `_AnisotropicPair` that needed to be exposed. I don't think they should be their own namespace. We could consider a submodule `md.pair.aniso.Dipole`.

[the remark this answers is not in the document]
Comment (joaander) [reply hoomd/md/pair.py:1653]: I was thinking we could use it to show that anisotropic potentials  all have `r_cut` and the shift modes like we do for `Pair`.

I like the idea of using `md.pair.aniso` for the potentials. It would make them stand out in the documentation and make it clear that they are different from the isotropic pair potentials. Let's do that.

File: hoomd/md/EvaluatorPairDipole.h (hunk 13 above)
Comment (b-butler) [reply hoomd/md/EvaluatorPairDipole.h]: Done.

File: hoomd/md/pair.py [hunk 15]
```diff
@@ -1784,78 +1693,73 @@ class gb(ai_pair):
...
 
-            * ``lpar`` (`float`, **required**) -  :math:`\ell_\parallel` (in distance units)
+            * ``lpar`` (`float`, **required**) -  :math:`\ell_\parallel` (in
+                distance units)
 
     Example::
 
         nl = nlist.Cell()
-        gb = pair.gb(nlist=nl, r_cut=2.5)
+        gb = md.pair.GB(nlist=nl, r_cut=2.5)
```
[the remark this answers is not in the document]
Comment (b-butler) [reply hoomd/md/pair.py]: Done.

**Review by b-butler — APPROVED** [review]
I made most of these recent changes, so I shouldn't really count as a reviewer.

File: hoomd/md/pair/aniso.py [hunk 16]
```diff
@@ -0,0 +1,222 @@
...
+
+    Attributes:
+        params (TypeParameter[tuple[``particle_type``, ``particle_type``], dict]):
+            The dipole potential parameters. The dictionary has the following
+            keys:
+
+            * ``A`` (`float`, **optional**) - :math:`A` - electrostatic energy
+              scale (*default*: 1.0)
+
+            * ``mu`` (`float`, **required**) - :math:`\mu` - emagnitude of
```
Comment (joaander) [inline hoomd/md/pair/aniso.py:93]: `mu` is not a key in the `params` TypeParameter, it is its own type parameter and accepts a full vector.

**Review by joaander — CHANGES_REQUESTED** [review]
Thanks, this looks great! Just one small fix to the docs and this is ready to merge.

**Comment by joaander** [comment]
In addition to the last documentation fix, you will need to merge master into this branch and resolve the conflicts.

**Review by joaander — APPROVED** [review]
Thanks!
